### ISSUE TITLE
OpenZFS not functional in RHEL9 after kernel update

### ISSUE THREAD

**Issue by pitrson** [openzfs/zfs#14698]
Hello folks, firstly system specs:
HW: ProLiant MicroServer Gen10 Plus
OS: RedHat 9.1
**Kernel**: 
- 5.14.0-162.22.2.el9_1.x86_64 (ZFS kernel modules can't be loaded (DKMS nor kABI) to kernel after update to this kernel)
- 5.14.0-162.12.1.el9_1.x86_64 (ZFS works as expected via kABI-tracking kmod, DKMS not tested)
Installation method:
Followed exactly all the steps mentioned in https://openzfs.github.io/openzfs-docs/Getting%20Started/RHEL-based%20distro/index.html.

kABI packages installed:
`[root@localhost extra]# rpm -qa | grep zfs
zfs-release-2-2.el9.noarch
libzfs5-2.1.9-1.el9.x86_64
kmod-zfs-2.1.9-1.el9.x86_64
zfs-2.1.9-1.el9.x86_64
`

**Issue**: after updating the kernel to 5.14.0-162.22.2.el9_1.x86_64 the ZFS kernel modules won't load. 

**Reproducing  / testing the issue:**
- install OpenZFS on RHEL 9.1 running 5.14.0-162.12.1.el9_1.x86_64 via kABI-tracking kmod
- update os/kernel in RHEL 9.1 as described above using standard method (dnf update) and reboot
- ZFS modules won't load (error: no such file or directory, previously seen modules in /lib/modules/<kernel version>/extra/ are not present)
- reinstallation of OpenZFS  (as per  https://openzfs.github.io/openzfs-docs/Getting%20Started/RHEL-based%20distro/index.html) doesn't lead to any change

**Notes:** 
After the above I have removed OpenZFS completely as per official documents and installed it via DKMS. After this installation the modules are visible in extra/ (xz compressed) but fail to load with following: "Invalid argument"

Note: I'm running the system with secure boot enabled, and I have signed even the DKMS generated modules, without any success (and it worked on the old kernel with kABI generated and signed modules)

- Am I doing something wrong or is that a bug which needs to be fixed? 
- Is this to be expected with every minor kernel update, or is it some major unexpected bug/behavior?  (if this is expected, I'll drop OpenZFS completely, even though this is "just" my homeserver. I have postponed this kernel update for few weeks now just to avoid such situations, yet it's not acceptable to me to work like this in 21st century, not even mentioning that RHEL is not using bleeding edge kernel)

If there are any debug logs/information required, I'll provide them upon your request.

Thanks!

**Comment by rincebrain** [comment]
`dmesg` output might be more detailed about what's broken. The CI doesn't apparently have RHEL, Alma, or any other derivative version 9 in it at this point - it used to, but that target was turned off in late October for some reason.

And RHEL isn't using a "bleeding edge" version, but they have a habit of backporting breaking changes from the bleeding edge sometimes, though not very often, and as ZFS requires and cannot avoid needing some symbols outside of the ones RHEL pinky promises won't break, this can break.

**Comment by pitrson** [comment]
Hi, 

there's nothing interesting in dmesg output. The current state after updating to new kernel -> reinstalling kABI -> removing kABI -> installing via DKMS -> removing DKMS -> installing via kABI again is:
`[root@localhost extra]# modprobe zfs
modprobe: ERROR: could not insert 'zfs': Invalid argument
`
`[root@localhost extra]# pwd
/lib/modules/5.14.0-162.22.2.el9_1.x86_64/extra
[root@localhost extra]# ls
avl  icp  icp.ko.xz  lua  nvpair  spl  spl.ko.xz  unicode  zavl.ko.xz  zcommon  zcommon.ko.xz  zfs  zfs.ko.xz  zlua.ko.xz  znvpair.ko.xz  zstd  zunicode.ko.xz  zzstd.ko.xz
`
Attached is  the strace of `modprobe zfs` maybe this can tell us more?
[strace_zfs.log](https://github.com/openzfs/zfs/files/11125241/strace_zfs.log)

**Comment by rincebrain** [comment]
No, that just tells us EINVAL.

Usually, the kernel logs more information about why it wouldn't load a thing after you try it and it says no.

**Comment by pitrson** [comment]
Hi there,

attached is the dmesg after fresh reboot. 
`[root@localhost ~]# systemctl --failed 
  UNIT                         LOAD   ACTIVE SUB    DESCRIPTION        
● systemd-modules-load.service loaded failed failed Load Kernel Modules
`
`Mar 31 22:01:17 localhost.localdomain systemd[1]: Starting Load Kernel Modules...
Mar 31 22:01:17 localhost.localdomain systemd-modules-load[5235]: Module 'msr' is built in
Mar 31 22:01:18 localhost.localdomain systemd-modules-load[5235]: Failed to insert module 'zfs': Invalid argument
Mar 31 22:01:18 localhost.localdomain systemd[1]: systemd-modules-load.service: Main process exited, code=exited, status=1/FAILURE
Mar 31 22:01:18 localhost.localdomain systemd[1]: systemd-modules-load.service: Failed with result 'exit-code'.
Mar 31 22:01:18 localhost.localdomain systemd[1]: Failed to start Load Kernel Modules.`

the `journalctl -xeu systemd-modules-load.service` also just says that the service is failed with `The process' exit code is 'exited' and its exit status is 1.` with no more details :( 
[dmesg.log](https://github.com/openzfs/zfs/files/11125551/dmesg.log)

**Comment by rincebrain** [comment]
![image](https://user-images.githubusercontent.com/214141/229245606-13a4b664-1302-4d64-ab19-1a6e03570b7b.png)

Seems to be working fine for me. I wonder what the difference is.

**Comment by pitrson** [comment]
Thanks!  I will give it a try this week, when I find a free spot ^^. 
I've installed kmod because that was the recommended way for "distribution" kernels. Should I rather stick to DKMS for the future?

**Comment by tonyhutter** [comment]
Just saw this issue - let me rebuild the kmods.

**Comment by pitrson** [comment]
Hi folks,

I have finally discovered what the problem was with DKMS (or with me? :-D )  (kmod doesn't work as per above updates currently).

The difference between kmod and dkms modules is, that openZFS DKMS modules are compressed with xz, kmod modules are not compressed at all. 

Since I'm running with Secureboot enabled, I have to sign such extra kernel modules. In previous kernel I've signed all the kmod modules manually (not compressed). Following the same procedure for DKMS didn't work, because if you sign the compressed module (xz file), it yields into "invalid argument" when trying to load such module (isn't it weird?).

However, if I uncompress it and sign, then compress it back again, all the modules load as expected! (thanks to https://unix.stackexchange.com/questions/438954/signing-a-compressed-kernel-module-for-use-with-secure-boot)

`for module in $(dirname $(modinfo -n zfs))/*.ko*; do
  module_basename=${module:0:-3}
  module_suffix=${module: -3}
  if [[ "$module_suffix" == ".xz" ]]; then
          unxz $module
          echo "signing $module"
          /usr/src/kernels/$(uname -r)/scripts/sign-file sha256 /etc/pki/pesign/sb_cert.priv /etc/pki/pesign/sb_cert.cer "${module_basename}"
          xz -f ${module_basename}
   else
        echo "Unrecognized suffix"
        exit 1
  fi
done

signing /lib/modules/5.14.0-162.22.2.el9_1.x86_64/extra/icp.ko.xz
signing /lib/modules/5.14.0-162.22.2.el9_1.x86_64/extra/spl.ko.xz
signing /lib/modules/5.14.0-162.22.2.el9_1.x86_64/extra/zavl.ko.xz
signing /lib/modules/5.14.0-162.22.2.el9_1.x86_64/extra/zcommon.ko.xz
signing /lib/modules/5.14.0-162.22.2.el9_1.x86_64/extra/zfs.ko.xz
signing /lib/modules/5.14.0-162.22.2.el9_1.x86_64/extra/zlua.ko.xz
signing /lib/modules/5.14.0-162.22.2.el9_1.x86_64/extra/znvpair.ko.xz
signing /lib/modules/5.14.0-162.22.2.el9_1.x86_64/extra/zunicode.ko.xz
signing /lib/modules/5.14.0-162.22.2.el9_1.x86_64/extra/zzstd.ko.xz
`

`[root@localhost extra]# modprobe zfs
[root@localhost extra]# lsmod | grep zfs
zfs                  3985408  0
zunicode              335872  1 zfs
zzstd                 581632  1 zfs
zlua                  192512  1 zfs
zavl                   16384  1 zfs
icp                   319488  1 zfs
zcommon               106496  2 zfs,icp
znvpair               114688  2 zfs,zcommon
spl                   122880  6 zfs,icp,zzstd,znvpair,zcommon,zavl
`

**Comment by pitrson** [comment]
Generally, I would then propose the following changes to the https://openzfs.github.io/openzfs-docs/Getting%20Started/RHEL-based%20distro/index.html# docu:

- state DKMS as preferred method of installation
- remove following : "DKMS packages are recommended for users running a non-distribution kernel or for users who wish to apply local customizations to OpenZFS. For most users the kABI-tracking kmod packages are recommended in order to avoid needing to rebuild OpenZFS for every kernel update." 
- add a small note: "secureboot enabled systems should ensure to uncompress -> sign -> compress DKMS modules ( as opposed to kmod where the modules are not compressed) "

I know that the solution is at the end related to knowledge on how to sign kernel modules, but the above small hints in the zfs docs would be helpful IMHO, especially in the use-cases I've reported. (I'll check whether I can submit a PR to the RHEL openZFS docs)

I'd also like to thank for your help, it's much appreciated!

**Comment by tonyhutter** [comment]
> Just saw this issue - let me rebuild the kmods.

kmods are released.  Try updating and you should see the zfs-2.1.9-2 kmod packages for EL9.

**Comment by tonyhutter** [comment]
I'm going to close this since the new kmods are out.  Please re-open if there are any issues.  Thanks all for being patient!

**Comment by pitrson** [comment]
one question @tonyhutter  - @rincebrain mentioned that "The CI doesn't apparently have RHEL, Alma, or any other derivative version 9 in it at this point - it used to, but that target was turned off in late October for some reason."

Does that mean that kmods are not automagically available for EL* shortly after new kernel is released ? Do you rebuild them only upon such reported issues? I'm just asking since I wanted to know whether I should really switch to DKMS instead of KMOD or not.

Thanks!

**Comment by tonyhutter** [comment]
@pitrson correct - the RPMs are not automatically generated when a new EL kernel drops.  The reason being that we need to manually sign the new RPMs & repos.

I would recommend that home users go with the dkms modules if they're running EL.

In the past, we used to have a RHEL builder that would act as a canary in the coal mine for whenever RedHat updated the kernel.  That would give us time to build new kmods whenever RHEL got updated, since CentOS usually lagged 2-4 weeks behind RHEL.   That's not really viable now though, as we're supporting AlmaLinux instead of CentOS, and Almalinux seems to be a lot faster about getting kernel updates to match RHEL.  We're also not running RHEL builders anymore.

To answer your other question - we're working on getting EL9 re-enabled in buildbot (https://github.com/openzfs/zfs-buildbot/pull/270)
